File: akka-sample-persistence-dc-java/src/main/java/sample/persistence/multidc/ThumbsUpSerializer.java
package sample.persistence.multidc;

import akka.actor.ExtendedActorSystem;
import akka.serialization.SerializerWithStringManifest;
import com.google.protobuf.InvalidProtocolBufferException;
import sample.persistence.multidc.protobuf.ThumbsUpMessages;

import java.io.NotSerializableException;
import java.util.HashSet;

public class ThumbsUpSerializer extends SerializerWithStringManifest {

  private final ExtendedActorSystem system;

  private final String StateManifest = "a";
  private final String GiveThumbsUpManifest = "b";
  private final String GetCountManifest = "c";
  private final String GetUsersManifest = "d";
  private final String GaveThumbsUpManifest = "e";

  public ThumbsUpSerializer(ExtendedActorSystem system) {
    this.system = system;
  }

  @Override
  public int identifier() {
    return 100;
  }

  @Override
  public String manifest(Object o) {
    if (o instanceof ThumbsUpCounter.GiveThumbsUp) return GiveThumbsUpManifest;
    else if (o instanceof ThumbsUpCounter.GaveThumbsUp) return GaveThumbsUpManifest;
    else if (o instanceof ThumbsUpCounter.State) return StateManifest;
    else if (o instanceof ThumbsUpCounter.GetCount) return GetCountManifest;
    else if (o instanceof ThumbsUpCounter.GetUsers) return GetUsersManifest;
    else
      throw new IllegalArgumentException("Can't serialize object of type " + o.getClass() + " in " + getClass().getName());
  }

  @Override
  public byte[] toBinary(Object o) {
    if (o instanceof ThumbsUpCounter.GiveThumbsUp) return giveThumbsUpToBinary((ThumbsUpCounter.GiveThumbsUp) o);
    else if (o instanceof ThumbsUpCounter.GaveThumbsUp) return gaveThumbsUpToBinary((ThumbsUpCounter.GaveThumbsUp) o);
    else if (o instanceof ThumbsUpCounter.State) return stateToBinary((ThumbsUpCounter.State) o);
    else if (o instanceof ThumbsUpCounter.GetCount) return getCountToBinary((ThumbsUpCounter.GetCount) o);
    else if (o instanceof ThumbsUpCounter.GetUsers) return getUsersToBinary((ThumbsUpCounter.GetUsers) o);
    else
      throw new IllegalArgumentException("Cannot serialize object of type " + o.getClass().getName());
  }

  private byte[] stateToBinary(ThumbsUpCounter.State a) {
    return ThumbsUpMessages.State.newBuilder()
        .addAllUsers(a.users)
        .build().toByteArray();
  }

  private byte[] giveThumbsUpToBinary(ThumbsUpCounter.GiveThumbsUp a) {
    return ThumbsUpMessages.GiveThumbsUp.newBuilder()
        .setResourceId(a.resourceId)
        .setUserId(a.userId)
        .build().toByteArray();
  }

  private byte[] getCountToBinary(ThumbsUpCounter.GetCount a) {
    return ThumbsUpMessages.GetCount.newBuilder()
        .setResourceId(a.resourceId)
        .build().toByteArray();
  }

  private byte[] getUsersToBinary(ThumbsUpCounter.GetUsers a) {
    return ThumbsUpMessages.GetUsers.newBuilder()
        .setResourceId(a.resourceId)
        .build().toByteArray();
  }

  private byte[] gaveThumbsUpToBinary(ThumbsUpCounter.GaveThumbsUp a) {
    return ThumbsUpMessages.GaveThumbsUp.newBuilder()
        .setUserId(a.userId)
        .build().toByteArray();
  }

  @Override
  public Object fromBinary(byte[] bytes, String manifest) throws NotSerializableException {
    try {
      if (manifest.equals(GiveThumbsUpManifest)) return giveThumbsUpFromBinary(bytes);
      else if (manifest.equals(GaveThumbsUpManifest)) return gaveThumbsUpFromBinary(bytes);
      else if (manifest.equals(StateManifest)) return stateFromBinary(bytes);
      else if (manifest.equals(GetCountManifest)) return getCountFromBinary(bytes);
      else if (manifest.equals(GetUsersManifest)) return getUsersFromBinary(bytes);
      else
        throw new NotSerializableException(
            "Unimplemented deserialization of message with manifest [" + manifest + "] in " + getClass().getName());
    } catch (InvalidProtocolBufferException e) {
      throw new NotSerializableException(e.getMessage());
    }
  }

  private ThumbsUpCounter.State stateFromBinary(byte[] bytes) throws InvalidProtocolBufferException {
    ThumbsUpMessages.State a = ThumbsUpMessages.State.parseFrom(bytes);
    return new ThumbsUpCounter.State(new HashSet<>(a.getUsersList()));
  }

  private ThumbsUpCounter.GiveThumbsUp giveThumbsUpFromBinary(byte[] bytes) throws InvalidProtocolBufferException {
    ThumbsUpMessages.GiveThumbsUp a = ThumbsUpMessages.GiveThumbsUp.parseFrom(bytes);
    return new ThumbsUpCounter.GiveThumbsUp(a.getResourceId(), a.getUserId());
  }

  private ThumbsUpCounter.GetCount getCountFromBinary(byte[] bytes) throws InvalidProtocolBufferException {
    ThumbsUpMessages.GetCount a = ThumbsUpMessages.GetCount.parseFrom(bytes);
    return new ThumbsUpCounter.GetCount(a.getResourceId());
  }

  private ThumbsUpCounter.GetUsers getUsersFromBinary(byte[] bytes) throws InvalidProtocolBufferException {
    ThumbsUpMessages.GetUsers a = ThumbsUpMessages.GetUsers.parseFrom(bytes);
    return new ThumbsUpCounter.GetUsers(a.getResourceId());
  }

  private ThumbsUpCounter.GaveThumbsUp gaveThumbsUpFromBinary(byte[] bytes) throws InvalidProtocolBufferException {
    ThumbsUpMessages.GaveThumbsUp a = ThumbsUpMessages.GaveThumbsUp.parseFrom(bytes);
    return new ThumbsUpCounter.GaveThumbsUp(a.getUserId());
  }

}

File: akka-sample-persistence-dc-scala/src/main/scala/sample/persistence/multidc/ThumbsUpSerializer.scala
package sample.persistence.multidc

import java.io.NotSerializableException

import akka.serialization.BaseSerializer
import akka.serialization.SerializerWithStringManifest
import sample.persistence.multidc.protobuf.ThumbsUpMessages
import scala.collection.JavaConverters._

class ThumbsUpSerializer(val system: akka.actor.ExtendedActorSystem)
  extends SerializerWithStringManifest with BaseSerializer {

  private val StateManifest = "a"
  private val GiveThumbsUpManifest = "b"
  private val GetCountManifest = "c"
  private val GetUsersManifest = "d"
  private val GaveThumbsUpManifest = "e"

  override def manifest(o: AnyRef): String = o match {
    case _: ThumbsUpCounter.GiveThumbsUp ⇒ GiveThumbsUpManifest
    case _: ThumbsUpCounter.GaveThumbsUp  ⇒ GaveThumbsUpManifest
    case _: ThumbsUpCounter.State   ⇒ StateManifest
    case _: ThumbsUpCounter.GetCount     ⇒ GetCountManifest
    case _: ThumbsUpCounter.GetUsers ⇒ GetUsersManifest

    case _ ⇒
      throw new IllegalArgumentException(s"Can't serialize object of type ${o.getClass} in [${getClass.getName}]")
  }

  override def toBinary(o: AnyRef): Array[Byte] = o match {
    case a: ThumbsUpCounter.GiveThumbsUp ⇒ giveThumbsUpToBinary(a)
    case a: ThumbsUpCounter.GaveThumbsUp  ⇒ gaveThumbsUpToBinary(a)
    case a: ThumbsUpCounter.State   ⇒ stateToBinary(a)
    case a: ThumbsUpCounter.GetCount     ⇒ getCountToBinary(a)
    case a: ThumbsUpCounter.GetUsers ⇒ getUsersToBinary(a)

    case _ ⇒
      throw new IllegalArgumentException(s"Cannot serialize object of type [${o.getClass.getName}]")
  }

  private def stateToBinary(a: ThumbsUpCounter.State): Array[Byte] = {
    val builder = ThumbsUpMessages.State.newBuilder()
    builder.addAllUsers(a.users.asJava)
    builder.build().toByteArray()
  }

  private def giveThumbsUpToBinary(a: ThumbsUpCounter.GiveThumbsUp): Array[Byte] = {
    giveThumbsUpToProto(a).build().toByteArray()
  }

  private def giveThumbsUpToProto(a: ThumbsUpCounter.GiveThumbsUp): ThumbsUpMessages.GiveThumbsUp.Builder = {
    val builder = ThumbsUpMessages.GiveThumbsUp.newBuilder()
    builder.setResourceId(a.resourceId).setUserId(a.userId)
    builder
  }

  private def getCountToBinary(a: ThumbsUpCounter.GetCount): Array[Byte] = {
    getCountToProto(a).build().toByteArray()
  }

  private def getCountToProto(a: ThumbsUpCounter.GetCount): ThumbsUpMessages.GetCount.Builder = {
    val builder = ThumbsUpMessages.GetCount.newBuilder()
    builder.setResourceId(a.resourceId)
    builder
  }

  private def getUsersToBinary(a: ThumbsUpCounter.GetUsers): Array[Byte] = {
    getUsersToProto(a).build().toByteArray()
  }

  private def getUsersToProto(a: ThumbsUpCounter.GetUsers): ThumbsUpMessages.GetUsers.Builder = {
    val builder = ThumbsUpMessages.GetUsers.newBuilder()
    builder.setResourceId(a.resourceId)
    builder
  }

  private def gaveThumbsUpToBinary(a: ThumbsUpCounter.GaveThumbsUp): Array[Byte] = {
    gaveThumbsUpToProto(a).build().toByteArray()
  }

  private def gaveThumbsUpToProto(a: ThumbsUpCounter.GaveThumbsUp): ThumbsUpMessages.GaveThumbsUp.Builder = {
    val builder = ThumbsUpMessages.GaveThumbsUp.newBuilder()
    builder.setUserId(a.userId)
    builder
  }

  override def fromBinary(bytes: Array[Byte], manifest: String): AnyRef = manifest match {
    case StateManifest   ⇒ stateFromBinary(bytes)
    case GiveThumbsUpManifest ⇒ giveThumbsUpFromBinary(bytes)
    case GaveThumbsUpManifest     ⇒ gaveThumbsUpFromBinary(bytes)
    case GetCountManifest ⇒ getCountFromBinary(bytes)
    case GetUsersManifest  ⇒ getUsersFromBinary(bytes)

    case _ ⇒
      throw new NotSerializableException(
        s"Unimplemented deserialization of message with manifest [$manifest] in [${getClass.getName}]")
  }

  private def stateFromBinary(bytes: Array[Byte]): ThumbsUpCounter.State = {
    val a = ThumbsUpMessages.State.parseFrom(bytes)
    ThumbsUpCounter.State(a.getUsersList.asScala.toSet)
  }

  private def giveThumbsUpFromBinary(bytes: Array[Byte]): ThumbsUpCounter.GiveThumbsUp = {
    val a = ThumbsUpMessages.GiveThumbsUp.parseFrom(bytes)
    ThumbsUpCounter.GiveThumbsUp(a.getResourceId, a.getUserId)
  }

  private def getCountFromBinary(bytes: Array[Byte]): ThumbsUpCounter.GetCount = {
    val a = ThumbsUpMessages.GetCount.parseFrom(bytes)
    ThumbsUpCounter.GetCount(a.getResourceId)
  }

  private def getUsersFromBinary(bytes: Array[Byte]): ThumbsUpCounter.GetUsers = {
    val a = ThumbsUpMessages.GetUsers.parseFrom(bytes)
    ThumbsUpCounter.GetUsers(a.getResourceId)
  }

  private def gaveThumbsUpFromBinary(bytes: Array[Byte]): ThumbsUpCounter.GaveThumbsUp = {
    val a = ThumbsUpMessages.GaveThumbsUp.parseFrom(bytes)
    ThumbsUpCounter.GaveThumbsUp(a.getUserId)
  }
}
